src/H5Tinit_float.c

```diff
@@ -0,0 +1,552 @@
+/*******************/
+
+/* Functions used in the DETECT_F() macro */
+static int          H5T__byte_cmp(int, const void *, const void *, const unsigned char *);
+static unsigned int H5T__bit_cmp(unsigned int, int *, void *, void *, const unsigned char *);
+static void         H5T__fix_order(int, int, int *, H5T_order_t *);
+static unsigned int H5T__imp_bit(unsigned int, int *, void *, void *, const unsigned char *);
+static unsigned int H5T__find_bias(unsigned int, unsigned int, int *, void *);
+static void         H5T__set_precision(H5T_fpoint_det_t *);
+
+/*-------------------------------------------------------------------------
+ * Function:    H5T__byte_cmp
+ *
+ * Purpose:     Compares two chunks of memory A and B and returns the
+ *              byte index into those arrays of the first byte that
+ *              differs between A and B.  Ignores differences where the
+ *              corresponding bit in pad_mask is set to 0.
+ *
+ * Return:      Success:    Index of differing byte.
+ *              Failure:    -1 if all bytes are the same.
+ *-------------------------------------------------------------------------
+ */
+static int
+H5T__byte_cmp(int n, const void *_a, const void *_b, const unsigned char *pad_mask)
+{
+    const unsigned char *a         = (const unsigned char *)_a;
+    const unsigned char *b         = (const unsigned char *)_b;
+    int                  ret_value = -1;
+
+    FUNC_ENTER_PACKAGE_NOERR
+
+    for (int i = 0; i < n; i++)
+        if ((a[i] & pad_mask[i]) != (b[i] & pad_mask[i]))
+            HGOTO_DONE(i);
+
+done:
+    FUNC_LEAVE_NOAPI(ret_value);
+}
+
+/*-------------------------------------------------------------------------
+ * Function:    H5T__bit_cmp
+ *
+ * Purpose:     Compares two bit vectors and returns the index for the
+ *              first bit that differs between the two vectors.     The
+ *              size of the vector is NBYTES.  PERM is a mapping from
+ *              actual order to little endian.  Ignores differences where
+ *              the corresponding bit in pad_mask is set to 0.
+ *
+ * Return:      Index of first differing bit.
+ *-------------------------------------------------------------------------
+ */
+static unsigned int
+H5T__bit_cmp(unsigned int nbytes, int *perm, void *_a, void *_b, const unsigned char *pad_mask)
+{
+    unsigned int   i;
+    unsigned char *a = (unsigned char *)_a;
+    unsigned char *b = (unsigned char *)_b;
+    unsigned char  aa, bb;
+    unsigned int   ret_value = 0;
+
+    FUNC_ENTER_PACKAGE_NOERR
+
+    for (i = 0; i < nbytes; i++) {
+        assert(perm[i] < (int)nbytes);
+        if ((aa = (unsigned char)(a[perm[i]] & pad_mask[perm[i]])) !=
+            (bb = (unsigned char)(b[perm[i]] & pad_mask[perm[i]]))) {
+            unsigned int j;
+
+            for (j = 0; j < 8; j++, aa >>= 1, bb >>= 1) {
+                if ((aa & 1) != (bb & 1))
+                    HGOTO_DONE(i * 8 + j);
+            }
+        }
+    }
```

Comment:
There should be some new error detection in this routine, to bring the previous error checking in.

Reply:
I can put the asserts back in, but they didn't seem particularly helpful.

Reply:
I think it would be better to return an error value and detect that in the calling code.